src/main/java/org/broadinstitute/consent/http/service/DatasetRegistrationService.java

```diff
@@ -0,0 +1,440 @@
+                                         User user,
+                                         String name,
+                                         FileCategory category) throws IOException {
+
+        FormDataBodyPart bodyPart = files.get(name);
+
+        if (!uploadedFileCache.containsKey(name)) {
+            BlobId id = uploadFile(bodyPart);
+            uploadedFileCache.put(name, id);
+        }
+
+        BlobId id = uploadedFileCache.get(name);
+
+        FileStorageObject fso = new FileStorageObject();
+        fso.setCategory(category);
+        fso.setFileName(bodyPart.getContentDisposition().getFileName());
+        fso.setMediaType(bodyPart.getMediaType().toString());
+        fso.setBlobId(id);
+        fso.setCreateUserId(user.getUserId());
+
+        return fso;
+    }
+
+
+    /**
+     * Extracts an individual field as a dataset property.
+     *
+     * @param name The human-readable name of the field
+     * @param schemaProp The schema property name (camelCase)
+     * @param type The type of the field, e.g. Boolean, String
+     * @param getField Lambda which gets the field's value
+     */
+    public record DatasetPropertyExtractor(
+            String name,
+            String schemaProp,
+            DatasetPropertyType type,
+            /*
+             * Takes in: Dataset registration object and consent group
+             * Produces: The value of the field, can be null if field not present.
+             */
+            BiFunction<DatasetRegistrationSchemaV1,ConsentGroup,Object> getField
+    ) {
+
+        /**
+         * Converts a field on the given registration to a DatasetProperty.
+         *
+         * @param registration The registration object to extract from
+         * @param consentGroup The index of the consent group to extract from
+         * @return The dataset property, if the field has a value, otherwise Optional.empty()
+         */
+        Optional<DatasetProperty> extract(DatasetRegistrationSchemaV1 registration, ConsentGroup consentGroup) {
+            Object value = this.getField.apply(registration, consentGroup);
+            if (Objects.isNull(value)) {
+                return Optional.empty();
+            }
+
+            DatasetProperty datasetProperty = new DatasetProperty();
+            datasetProperty.setPropertyName(this.name);
+            datasetProperty.setPropertyType(this.type);
+            datasetProperty.setSchemaProperty(this.schemaProp);
+            datasetProperty.setPropertyValue(this.type.coerce(value.toString()));
+
+            return Optional.of(datasetProperty);
+
+        }
+    };
+
+    private static final List<DatasetPropertyExtractor> DATASET_REGISTRATION_V1_PROPERTY_EXTRACTORS = List.of(
+            new DatasetPropertyExtractor(
+                    "PI Name", "piName", DatasetPropertyType.String,
+                    (registration, consentGroup) -> registration.getPiName()),
```

Comment:
Once again I wonder if there's a better way to do this in an automatic way so we don't miss anything.

Reply:
> Once again I wonder if there's a better way to do this in an automatic way so we don't miss anything.

I have the same concern. `"piName"` is used in several places, for example. We don't have anything that, say, correlates the two strings, `"PI Name"` and `"piName"` however. Since these come from `DatasetRegistrationSchemaV1`, I'm not seeing a clear solution to aligning all of them. The other complicating factor here is that `DatasetRegistrationSchemaV1` is made with a code generator so maintaining the original json schema with manual changes to that class is challenging.